riscv/src/compiler.rs

```diff
@@ -520,7 +526,10 @@ fn preamble(degree: u64, coprocessors: &CoProcessors, with_bootloader: bool) ->
 
     // Except for the last row, if m_change is 1, then m_addr has to increase,
     // if it is zero, m_step has to increase.
-    (1 - LAST) { m_change * (m_addr' - m_addr) + (1 - m_change) * (m_step' - m_step) } in POSITIVE;
+    // `m_diff_upper * 2**16 + m_diff_lower` has to be equal to the difference **minus one**.
+    // Since we know that both m_addr and m_step can only be 32-Bit, this enforces that
+    // the values are strictly increasing.
+    (1 - LAST) * (m_change * (m_addr' - m_addr) + (1 - m_change) * (m_step' - m_step) - 1 - m_diff_upper * 2**16 - m_diff_lower) = 0;
```

Comment:
can you change this to use intermediate columns and keep the degree of each constraint 2?